#6024 New Feature - Relate/Assign Incoming Email

ApatheticCosmos opened the pull request:
Change how IncomingEmail handles assigning and relating email retrieved from group mail account.
See the thread at https://suitecrm.com/suitecrm/forum/suitecrm-7-0-discussion/19458-developing-incoming-email-features-request-for-comment for more info and a screenshot of the configuration option.
Issue #6024
<!--- Provide a general summary of your changes in the Title above -->

## Description
With this patch, all email replies retrieved from a Group Mail account is related with the same record as referenced sent email.
It also provides the option to assign the email to either the owner of the referenced email, or to the owner of the record related with the referenced email.
It sets the message_id on all outgoing emails. See Issue #5977.

<!--- Describe your changes in detail -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here unless your commit contains the issue number -->

## Motivation and Context
Currently, SuiteCRM has no idea what record or user a group incoming email is related with, or is in reply to. The only information is has to go on is the from email address, which must match an email address in a record to even show up there. It currently can't set the parent record of the incoming email properly.
Also, the only way to change ownership of an incoming group email is with workflows, or manually. With this patch, incoming email can be automatically assigned to the user that owns the parent record of the thread, or the user that sent the referenced email.
<!--- Why is this change required? What problem does it solve? -->

## How To Test This
When you apply this patch, all outgoing email will get the mesaage_id set in the emails table.
Any email that is retrieved by a group account and references that message_id will be related to the record of the original email. You can also tell it to assign to the owner of that record, or the owner of the referenced email, or do nothing different than before.
<!--- Please describe in detail how to test your changes. -->

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Final checklist
<!--- Go over all the following points and check all the boxes that apply. --->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! --->
- [X] My code follows the code style of this project found [here](https://docs.suitecrm.com/community/contributing-code/coding-standards/).
- [X] My change requires a change to the documentation.
    (I think it does, although it's documented in the help tip)
- [X] I have read the [**How to Contribute**](https://docs.suitecrm.com/community/contributing-code/) guidelines.

<!--- Your pull request will be tested via Travis CI to automatically indicate that your changes do not prevent compilation. --->

<!--- If it reports back that there are problems, you can log into the Travis system and check the log report for your pull request to see what the problem was. --->

isleshocky77 committed:
Adding the abily to relate a Note with a Campaign

isleshocky77 committed:
PSR-2 Convert tabs to spaces

isleshocky77 committed:
Code Format after tab -> spaces conversion

isleshocky77 committed:
Remove duplicate array key (clean warning)

OffsiteGuru committed:
Feature - Associate Incoming Email (#1)

Change how IncomingEmail handles assigning and associating email retrieved from group mail account.

PedroErnst commented on modules/InboundEmail/InboundEmail.php:
```diff
@@ -5302,7 +5302,39 @@ public function returnImportedEmail($msgNo, $uid, $forDisplay = false, $clean_em
+                                        break;
+                                    case 'recordowner'; // Assign the email to the owner of the record the referenced email's parrent
+                                        $parentBean = BeanFactory::getBean($row['parent_type'], $row['parent_id']);
+                                        $email->assigned_user_id = $parentBean->assigned_user_id;
```
Small check here to make sure $parentBean isn't null could help prevent PHP notices.

ApatheticCosmos replied:
Done!

PedroErnst reviewed (dismissed):
Very nice feature to have!

OffsiteGuru committed:
Suggestion of PedroErnst

Check to make sure the parentBean has a non-null id.

labcerouno commented:
After applying this fix and configuring a new mail account, I get the following FATAL ERROR:

MySQL error 1054: Unknown column 'assignment_behavior' in 'field list'

It's quite obvious the reason. So I tried adding this field in the InboundEmail table and it seems to be working fine!

gymad reviewed (commented):
I think as it is a new feature, it should be going to the `develop` branch. Can you please change the target branch, please

Mausino commented:
@gymad should the issue to which is this pull request assigned label by milestone 7.11 ? 
This is very usefull feature which save lots of time

ApatheticCosmos commented:
Can we get the 'wrong branch' tag removed from this please?

samus-aran commented:
Thanks @ApatheticCosmos We are circulating this internally for review :+1:

Dillon-Brown committed:
Remove fullcalendar

Dillon-Brown committed:
Fixed #3057 - FullCalendar update to version 3.1

Dillon-Brown committed:
Fixed #4511 - Upgrade of TinyMCE to latest 4.7.x version

Dillon-Brown committed:
Remove tinyMCE

Dillon-Brown committed:
Fixed #3100 - Upgrade for HTML Purifier [Request]

gymad commented on modules/Emails/EmailsController.php:
```diff
@@ -171,9 +171,13 @@ public function action_send()
         $request = $_REQUEST;
 
+        // Set message_id for email record in DB
+        $request['message_id'] = 'SUITECRM_' . md5('HELLO'.(idate("U")-1000000000).uniqid()) . '@' . $sugar_config['host_name'];
```
Calling uniqid() without any parameter is a bit dangerous ans not a best option as it's related on a system time clock.
see more in php.net: 
> Warning
This function does not guarantee uniqueness of return value. Since most systems adjust system clock by NTP or like, system time is changed constantly. Therefore, it is possible that this function does not return unique ID for the process/thread. Use more_entropy to increase likelihood of uniqueness.
http://php.net/manual/en/function.uniqid.php#120123

laszlocsontos replied:
I'm trying to implement open and click tracking with this upcoming feature.

My use case is that I'm using Mailgun that provides callbacks when these emails are opened. Mailgun refers back to emails which were acted upon by their `Message-Id`. If the hostname isn't part of that value which is saved to the DB, an additional processing step is needed to split that value by `@` and do the lookup based on the first part.

From this perspective, it would be more straightforward, if the value of the `message_id` field in the DB was exactly the same as the value of the `Message-Id` field in sent emails.

Currently I'm using this in my custom SuiteCRM installation:

```
$request['message_id'] = '<' . time().'.'.bin2hex(random_bytes(8)).'@'.gethostname().'>';
```

This produces a value, something like `<1549445373.02231e5c88e1a743@typhoon>`. 

What do you think @ApatheticCosmos @gymad?

gymad commented on modules/InboundEmail/InboundEmail.php:
```diff
@@ -5302,7 +5302,41 @@ public function returnImportedEmail($msgNo, $uid, $forDisplay = false, $clean_em
                 }
                 $email->$rel->add($email->parent_id);
-            }
+            } else {
```
This function complexity is getting hardly, almost unreadable. Please split out into separated functions.

gymad commented on modules/InboundEmail/InboundEmail.php:
```diff
@@ -5302,7 +5302,41 @@ public function returnImportedEmail($msgNo, $uid, $forDisplay = false, $clean_em
+                    foreach ($references as $reference) {
+                        $referenceId = rtrim(ltrim($reference, '<'), '>');
+                        if (strncmp($referenceId, "SUITECRM_", 9) === 0) {
+                            $query = 'SELECT id, assigned_user_id, parent_id, parent_type FROM emails WHERE emails.message_id = \'' . $referenceId . '\' and type = \'out\' and emails.deleted = 0 LIMIT 1';
```
Use quote() or quoted() on query parameters.

gymad commented on modules/InboundEmail/InboundEmail.php:
```diff
@@ -5302,7 +5302,41 @@ public function returnImportedEmail($msgNo, $uid, $forDisplay = false, $clean_em
+                                        break;
+                                    case 'recordowner'; // Assign the email to the owner of the record the referenced email's parrent
+                                        $parentBean = BeanFactory::getBean($row['parent_type'], $row['parent_id']);
+                                        if (isset($parentBean->id)) {
```
Testing for $parentBean->id but using $parentBean->assigned_user_id. Is it intend?

gymad commented on modules/InboundEmail/InboundEmail.php:
```diff
@@ -5302,7 +5302,41 @@ public function returnImportedEmail($msgNo, $uid, $forDisplay = false, $clean_em
+                                            $email->assigned_user_id = $parentBean->assigned_user_id;
+                                        }
+                                        break;
+                                    case 'repliedetoowner'; // Assign the email to the owner of the referenced email
```
Mispelled. Did you test it before made this PR?

gymad commented on modules/InboundEmail/InboundEmail.php:
```diff
@@ -5302,7 +5302,41 @@ public function returnImportedEmail($msgNo, $uid, $forDisplay = false, $clean_em
+                                        $email->assigned_user_id = $row['assigned_user_id'];
+                                        break;
+                                    default;
+                                        $GLOBALS['log']->error('Inbound Email misconfiguration. Assignment Behavior set to invalid value. Not assigning email.');
```
Use LoggerManager instead superglobals.

gymad commented on modules/InboundEmail/InboundEmail.php:
```diff
@@ -5302,7 +5302,41 @@ public function returnImportedEmail($msgNo, $uid, $forDisplay = false, $clean_em
+                        } // END if (strncmp($referenceId, "SUITECRM_", 9) === 0)
+                    } // END foreach ($references as $reference)
+                } // END if(!empty($header->references) || !empty($header->in_reply_to))
+            } // END else
```
Tons of IF but no ELSE. If I will call this function how can I determ what happend in this area?

Dillon-Brown committed:
Merge pull request #6714 from gymad/develop_elasticpager_fix

fix last page when modulus is zero

Dillon-Brown committed:
Merge branch 'develop' into tiny_mce

Dillon-Brown committed:
Merge pull request #6730 from Dillon-Brown/issue-6720

Fixed #6720 - Remove old recaptcha files

Dillon-Brown committed:
Merge branch 'develop' into fullcalendar

Dillon-Brown committed:
Merge branch 'develop' into fullcalendar

isleshocky77 committed:
Cleaning up the clean cache command

* Update command to allow for cache directory specified in config
* Remove warnings/errors when no cached directory present
* Update to have namespace "cache" so we can add other cache commands later
  * Leave alias for current command
* Use provided toolsets required by robo for implementation
* Print cleaned directories
* Confirm before cleaning directory
  * Allow for force override

Dillon-Brown committed:
Merge branch 'develop' into fix-clean-cache-command-config-cache-dir

isleshocky77 committed:
Code formatting

Dillon-Brown committed:
Merge remote-tracking branch 'upstream/develop' into feature/campaign-notes

# Conflicts:
#	modules/Campaigns/Campaign.php
#	modules/Campaigns/views/view.detail.php
#	modules/Notes/vardefs.php

Dillon-Brown committed:
Reformat

Dillon-Brown committed:
Fix test

samus-aran committed:
Merge pull request #4606 from isleshocky77/feature/campaign-notes

Add ability to have Notes on a Campaign

Dillon-Brown committed:
Formatting

Dillon-Brown committed:
Merge pull request #6802 from isleshocky77/fix-clean-cache-command-config-cache-dir

Allow for clean cache command to use config'd cache_dir (plus clean up)

Dillon-Brown committed:
Remove PHP4 constructors

Dillon-Brown committed:
Nusoap PHP4 style constructors

Dillon-Brown committed:
More PHP4 style constructors

Dillon-Brown committed:
Review comments

cherrador committed:
Update modules/AOR_Charts/lib/pChart/class/pRadar.class.php

Co-Authored-By: Dillon-Brown <26431166+Dillon-Brown@users.noreply.github.com>

lazka committed:
Fixes #7111 - composer: include mpdf/mpdf 6.1 and drop in-tree version (5.7)

Remove the included version of mpdf (in modules/AOS_PDF_Templates/PDF_Lib) and
add it to composer instead.

This also updates the version from 5.7.1 to 6.1.0 to get some fixes that were already
applied manually to the version in SuiteCRM. There are many newer versions available,
but they require more changes and this is a good first step (and this potentially
could be backported to 7.10)

The version in tree had one manual fix applied (d1279cad9c717) which is included in
6.1.0 already, so the upstream version can be used as is.

The required changes to make things compatible with with 6.1 were
* mPDF_exception is now called MpdfException
* SetAutoFont() no longer exists, the docs suggest to use autoLangToFont instead
  to get the same results as before.

Fixes #7111

Dillon-Brown committed:
Merge remote-tracking branch 'upstream/develop' into fullcalendar

# Conflicts:
#	composer.json
#	composer.lock

Dillon-Brown committed:
Merge remote-tracking branch 'upstream/hotfix' into develop_merge

code-ph0y committed:
Merge pull request #6702 from Dillon-Brown/fullcalendar

Fixed #3057 - FullCalendar update to version 3.1

Dillon-Brown committed:
Merge branch 'develop' into develop_merge

samus-aran committed:
Merge pull request #6703 from Dillon-Brown/htmlpurifier

Fixed #3100 - Upgrade for HTML Purifier [Request]

Dillon-Brown committed:
Merge remote-tracking branch 'upstream/master' into develop_merge

Dillon-Brown committed:
Merge remote-tracking branch 'origin/develop_merge' into develop_merge

# Conflicts:
#	composer.json

Dillon-Brown committed:
Update composer.lock

Dillon-Brown committed:
Merge pull request #7131 from Dillon-Brown/develop_merge

Develop merge

Dillon-Brown committed:
Merge branch 'develop' into remove_constructors

Dillon-Brown committed:
Add pclzip/pclzip to composer

Dillon-Brown committed:
Remove pclzip references

Dillon-Brown committed:
Remove pclZip from include

Dillon-Brown committed:
Update composer.lock

Dillon-Brown committed:
Remove zip_utils.php

Dillon-Brown committed:
Reformat php_zip_utils.php

Dillon-Brown committed:
Update composer

Dillon-Brown committed:
Merge remote-tracking branch 'upstream/develop' into tiny_mce

# Conflicts:
#	composer.json
#	composer.lock

Dillon-Brown committed:
Update composer

Dillon-Brown committed:
Update tinymce references

Dillon-Brown committed:
Add tedivm/jshrink to composer

Dillon-Brown committed:
Remove Minifier.php

Dillon-Brown committed:
Fixed #7027 - Use a maintained library to minify JS

Dillon-Brown committed:
Merge remote-tracking branch 'upstream/develop' into fix-mpdf-composer

# Conflicts:
#	composer.json
#	composer.lock

Dillon-Brown committed:
Update composer

Dillon-Brown committed:
PHP 5.6 compat

code-ph0y committed:
Merge pull request #6699 from Dillon-Brown/tiny_mce

Fixed #4511 - Upgrade of TinyMCE to latest 4.9.x version

Dillon-Brown committed:
Merge branch 'develop' into fix-mpdf-composer

Dillon-Brown committed:
Merge pull request #7118 from lazka/fix-mpdf-composer

Fixes #7111 - composer: include mpdf/mpdf 6.1 and drop in-tree version (5.7)

code-ph0y committed:
Merge branch 'develop' into issue-7027

cameronblaikie committed:
Merge pull request #7295 from Dillon-Brown/issue-7027

Fixed #7027 - Use a maintained library to minify JS

Dillon-Brown committed:
Merge remote-tracking branch 'upstream/develop' into feature/plcZip

# Conflicts:
#	composer.json
#	composer.lock

Dillon-Brown committed:
Update composer.lock

Dillon-Brown committed:
Merge remote-tracking branch 'upstream/develop' into feature/plcZip

# Conflicts:
#	composer.lock

Dillon-Brown committed:
Update composer.lock

Dillon-Brown committed:
Merge remote-tracking branch 'upstream/develop' into remove_constructors

# Conflicts:
#	modules/AOS_PDF_Templates/PDF_Lib/classes/bmp.php
#	modules/AOS_PDF_Templates/PDF_Lib/classes/cssmgr.php
#	modules/AOS_PDF_Templates/PDF_Lib/classes/directw.php
#	modules/AOS_PDF_Templates/PDF_Lib/classes/form.php
#	modules/AOS_PDF_Templates/PDF_Lib/classes/gif.php
#	modules/AOS_PDF_Templates/PDF_Lib/classes/grad.php
#	modules/AOS_PDF_Templates/PDF_Lib/classes/indic.php
#	modules/AOS_PDF_Templates/PDF_Lib/classes/svg.php
#	modules/AOS_PDF_Templates/PDF_Lib/classes/tocontents.php
#	modules/AOS_PDF_Templates/PDF_Lib/classes/ttfontsuni.php
#	modules/AOS_PDF_Templates/PDF_Lib/classes/wmf.php
#	modules/AOS_PDF_Templates/PDF_Lib/mpdf.php
#	modules/AOS_PDF_Templates/PDF_Lib/mpdfi/fpdi_pdf_parser.php
#	modules/AOS_PDF_Templates/PDF_Lib/mpdfi/pdf_context.php
#	modules/AOS_PDF_Templates/PDF_Lib/mpdfi/pdf_parser.php

Dillon-Brown committed:
Merge pull request #7287 from Dillon-Brown/feature/plcZip

Remove PclZip

connorshea committed:
Remove HTMLPurifier ConfigForm.js file.

This probably isn't necessary and can be removed?

Dillon-Brown committed:
Merge remote-tracking branch 'upstream/hotfix' into develop_merge_1

# Conflicts:
#	composer.lock
#	files.md5
#	include/HTMLPurifier/HTMLPurifier.standalone.php
#	include/HTMLPurifier/standalone/HTMLPurifier/Lexer/PH5P.php
#	include/SugarCharts/Jit/FlashCanvas/canvas2png.js
#	include/SugarCharts/Jit/FlashCanvas/flashcanvas.js
#	include/SugarCharts/Jit/js/Jit/jit.js
#	include/SugarCharts/Jit/js/sugarCharts.js
#	include/pclzip/pclzip.lib.php
#	include/utils/zip_utils.php
#	lib/Robo/Plugin/Commands/CleanCacheCommands.php
#	modules/AOS_PDF_Templates/PDF_Lib/classes/barcode.php
#	modules/AOS_PDF_Templates/PDF_Lib/classes/bmp.php
#	modules/AOS_PDF_Templates/PDF_Lib/classes/cssmgr.php
#	modules/AOS_PDF_Templates/PDF_Lib/classes/directw.php
#	modules/AOS_PDF_Templates/PDF_Lib/classes/form.php
#	modules/AOS_PDF_Templates/PDF_Lib/classes/gif.php
#	modules/AOS_PDF_Templates/PDF_Lib/classes/grad.php
#	modules/AOS_PDF_Templates/PDF_Lib/classes/indic.php
#	modules/AOS_PDF_Templates/PDF_Lib/classes/meter.php
#	modules/AOS_PDF_Templates/PDF_Lib/classes/svg.php
#	modules/AOS_PDF_Templates/PDF_Lib/classes/tocontents.php
#	modules/AOS_PDF_Templates/PDF_Lib/classes/ttfontsuni.php
#	modules/AOS_PDF_Templates/PDF_Lib/classes/ttfontsuni_analysis.php
#	modules/AOS_PDF_Templates/PDF_Lib/classes/wmf.php
#	modules/AOS_PDF_Templates/PDF_Lib/compress.php
#	modules/AOS_PDF_Templates/PDF_Lib/config.php
#	modules/AOS_PDF_Templates/PDF_Lib/config_cp.php
#	modules/AOS_PDF_Templates/PDF_Lib/config_fonts.php
#	modules/AOS_PDF_Templates/PDF_Lib/font/ccourier.php
#	modules/AOS_PDF_Templates/PDF_Lib/font/ccourierb.php
#	modules/AOS_PDF_Templates/PDF_Lib/font/ccourierbi.php
#	modules/AOS_PDF_Templates/PDF_Lib/font/ccourieri.php
#	modules/AOS_PDF_Templates/PDF_Lib/font/chelvetica.php
#	modules/AOS_PDF_Templates/PDF_Lib/font/chelveticab.php
#	modules/AOS_PDF_Templates/PDF_Lib/font/chelveticabi.php
#	modules/AOS_PDF_Templates/PDF_Lib/font/chelveticai.php
#	modules/AOS_PDF_Templates/PDF_Lib/font/csymbol.php
#	modules/AOS_PDF_Templates/PDF_Lib/font/ctimes.php
#	modules/AOS_PDF_Templates/PDF_Lib/font/ctimesb.php
#	modules/AOS_PDF_Templates/PDF_Lib/font/ctimesbi.php
#	modules/AOS_PDF_Templates/PDF_Lib/font/ctimesi.php
#	modules/AOS_PDF_Templates/PDF_Lib/font/czapfdingbats.php
#	modules/AOS_PDF_Templates/PDF_Lib/graph.php
#	modules/AOS_PDF_Templates/PDF_Lib/includes/CJKdata.php
#	modules/AOS_PDF_Templates/PDF_Lib/includes/functions.php
#	modules/AOS_PDF_Templates/PDF_Lib/includes/ind_bn_1_001.volt.php
#	modules/AOS_PDF_Templates/PDF_Lib/includes/ind_gu_1_001.volt.php
#	modules/AOS_PDF_Templates/PDF_Lib/includes/ind_hi_1_001.volt.php
#	modules/AOS_PDF_Templates/PDF_Lib/includes/ind_kn_1_001.volt.php
#	modules/AOS_PDF_Templates/PDF_Lib/includes/ind_ml_1_001.volt.php
#	modules/AOS_PDF_Templates/PDF_Lib/includes/ind_or_1_001.volt.php
#	modules/AOS_PDF_Templates/PDF_Lib/includes/ind_pa_1_001.volt.php
#	modules/AOS_PDF_Templates/PDF_Lib/includes/ind_ta_1_001.volt.php
#	modules/AOS_PDF_Templates/PDF_Lib/includes/ind_te_1_001.volt.php
#	modules/AOS_PDF_Templates/PDF_Lib/includes/out.php
#	modules/AOS_PDF_Templates/PDF_Lib/includes/subs_core.php
#	modules/AOS_PDF_Templates/PDF_Lib/includes/subs_win-1252.php
#	modules/AOS_PDF_Templates/PDF_Lib/includes/upperCase.php
#	modules/AOS_PDF_Templates/PDF_Lib/mpdf.php
#	modules/AOS_PDF_Templates/PDF_Lib/mpdfi/filters/FilterASCII85.php
#	modules/AOS_PDF_Templates/PDF_Lib/mpdfi/filters/FilterLZW.php
#	modules/AOS_PDF_Templates/PDF_Lib/mpdfi/fpdi_pdf_parser.php
#	modules/AOS_PDF_Templates/PDF_Lib/mpdfi/pdf_context.php
#	modules/AOS_PDF_Templates/PDF_Lib/mpdfi/pdf_parser.php
#	modules/AOS_PDF_Templates/PDF_Lib/patterns/de.php
#	modules/AOS_PDF_Templates/PDF_Lib/patterns/en.php
#	modules/AOS_PDF_Templates/PDF_Lib/patterns/es.php
#	modules/AOS_PDF_Templates/PDF_Lib/patterns/fi.php
#	modules/AOS_PDF_Templates/PDF_Lib/patterns/fr.php
#	modules/AOS_PDF_Templates/PDF_Lib/patterns/it.php
#	modules/AOS_PDF_Templates/PDF_Lib/patterns/nl.php
#	modules/AOS_PDF_Templates/PDF_Lib/patterns/pl.php
#	modules/AOS_PDF_Templates/PDF_Lib/patterns/sv.php
#	modules/AOS_PDF_Templates/PDF_Lib/utils/UnicodeRanges.php
#	modules/AOS_PDF_Templates/PDF_Lib/utils/font_collections.php
#	modules/AOS_PDF_Templates/PDF_Lib/utils/font_coverage.php
#	modules/AOS_PDF_Templates/PDF_Lib/utils/font_dump.php
#	modules/AOS_PDF_Templates/PDF_Lib/utils/font_names.php
#	modules/Emails/include/ListView/DeleteEmailAction.js
#	modules/UpgradeWizard/silentUpgrade_step1.php

Dillon-Brown committed:
Update composer.lock

Mac-Rae commented:
Please resolve the conflicts on this pull request bringing it up to date.

samus-aran committed:
Merge pull request #7674 from Dillon-Brown/develop_merge_1

Develop merge

Dillon-Brown committed:
Merge remote-tracking branch 'upstream/develop' into remove_constructors

Dillon-Brown committed:
Merge remote-tracking branch 'upstream/develop' into remove_constructors

# Conflicts:
#	include/nusoap/class.soap_parser.php
#	include/nusoap/class.xmlschema.php

samus-aran committed:
Merge pull request #6750 from Dillon-Brown/remove_constructors

Remove PHP4 style constructors

Dillon-Brown committed:
Merge pull request #7606 from connorshea/remove-orphaned-html-purifier-file

Remove HTMLPurifier ConfigForm.js file.

ApatheticCosmos committed:
Merge branch 'develop' into feature/associate_incoming_email

ApatheticCosmos commented:
@Mac-Rae Conflict Resolved. I'm going to be doing some more work to clean up the things @gymad has pointed out as well.

serhiisamko091184 commented:
Hello, @ApatheticCosmos 

The PR has been marked as stale because there has been no recent activity. It will be closed if no further activity occurs.
Thanks for your contributions.

Regards,
Serhii

serhiisamko091184 commented:
Hey @ApatheticCosmos 

Closing this PR due to inactivity. 
Thank you.

Regards,
Serhii
